stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/avatar/AvatarBitmapFactory.kt

```diff
@@ -0,0 +1,135 @@
+package io.getstream.chat.android.ui.avatar
+
+import android.content.Context
+import android.graphics.Bitmap
+import android.graphics.Canvas
+import android.graphics.LinearGradient
+import android.graphics.Paint
+import android.graphics.Shader
+import android.graphics.Typeface
+import android.graphics.drawable.BitmapDrawable
+import coil.Coil
+import coil.request.ImageRequest
+import com.getstream.sdk.chat.utils.LlcMigrationUtils
+import io.getstream.chat.android.client.models.Channel
+import io.getstream.chat.android.client.models.User
+import io.getstream.chat.android.client.models.image
+import io.getstream.chat.android.ui.R
+import io.getstream.chat.android.ui.avatar.AvatarView.Companion.MAX_AVATAR_SECTIONS
+import io.getstream.chat.android.ui.utils.adjustColorLBrightness
+import io.getstream.chat.android.ui.utils.getIntArray
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.withContext
+
+internal class AvatarBitmapFactory(private val context: Context) {
+    private val gradientBaseColors = context.getIntArray(R.array.avatar_gradient_colors)
+
+    internal suspend fun createUserBitmap(user: User, style: AvatarStyle): Bitmap {
+        return createBitmapFromUrl(user.image)
+            ?: createInitialsBitmap(style, LlcMigrationUtils.getInitials(user))
+    }
+
+    internal suspend fun createChannelBitmaps(
+        channel: Channel,
+        lastActiveUsers: List<User>,
+        style: AvatarStyle
+    ): List<Bitmap> {
+        return createBitmapFromUrl(channel.image)
+            ?.let { listOf(it) }
+            ?: createUsersBitmaps(lastActiveUsers, style).takeUnless { it.isEmpty() }
+            ?: listOf(createInitialsBitmap(style, LlcMigrationUtils.getInitials(channel)))
+    }
+
+    private suspend fun createUsersBitmaps(users: List<User>, style: AvatarStyle): List<Bitmap> {
+        return users.take(MAX_AVATAR_SECTIONS).map { createUserBitmap(it, style) }
+    }
+
+    private suspend fun createBitmapFromUrl(
+        url: String
+    ): Bitmap? = withContext(Dispatchers.IO) {
+        url.takeUnless { it.isBlank() }
+            ?.let {
+                (
+                    Coil.execute(
+                        ImageRequest.Builder(context)
+                            .data(it)
+                            .allowHardware(false)
```

Comment:
Why are we disabling it?

Reply:
I got some crashes as the bitmap was mutated which is not allowed in hardware mode. But now it works fine so I removed `allowHardware(false)`. Actually just utilized `ImageLoader.getBitmap()` where hardware mode is enabled.